File: Java/1146. Snapshot Array.java
class SnapshotArray {
    HashMap<Integer, Integer>[] snaps;
    int id;
    
    public SnapshotArray(int length) {
        snaps = new HashMap[length];
        for(int i=0; i<length; i++) {
            snaps[i] = new HashMap<>();
        }
        id = 0;
    }
    
    // map the version with value
    public void set(int index, int val) {
        snaps[index].put(id, val);
    }
    
    public int snap() {
        id++;
        return id-1;
    }
    
    public int get(int index, int snap_id) {
        while(snap_id >= 0) {
            if(snaps[index].containsKey(snap_id)) {
                return snaps[index].get(snap_id);
            } else {
                snap_id--;
            }
        }
        return 0;
    }
}

/**
 * Your SnapshotArray object will be instantiated and called as such:
 * SnapshotArray obj = new SnapshotArray(length);
 * obj.set(index,val);
 * int param_2 = obj.snap();
 * int param_3 = obj.get(index,snap_id);
 */

File: Python3/1146. Snapshot Array.py
class SnapshotArray:

    def __init__(self, length: int):
        self.snapshot = {0: {}}
        self.id = 0

    def set(self, index: int, val: int) -> None:
        self.snapshot[self.id][index] = val

    def snap(self) -> int:
        self.snapshot[self.id+1] = self.snapshot[self.id].copy()
        self.id += 1
        return self.id-1

    def get(self, index: int, snap_id: int) -> int:
        return self.snapshot[snap_id][index] if index in self.snapshot[snap_id] else 0


# Your SnapshotArray object will be instantiated and called as such:
# obj = SnapshotArray(length)
# obj.set(index,val)
# param_2 = obj.snap()
# param_3 = obj.get(index,snap_id)